python/ray/util/sgd/torch/torch_dataset.py

```diff
@@ -0,0 +1,138 @@
+import functools
+import logging
+from collections import Iterator
+from collections.abc import Iterable
+from typing import Any, Callable, List, Optional
+
+import numpy as np
+import torch
+from pandas import DataFrame
+from torch.utils.data import IterableDataset
+
+from ray.util.data.dataset import MLDataset
+
+
+def convert_to_tensor(df, feature_columns: List[str],
+                      feature_shapes: List[Any],
+                      feature_types: List[torch.dtype], label_column: str,
+                      label_shape: Optional[int], label_type: torch.dtype):
+    feature_tensor = []
+    for col, shape, dtype in zip(feature_columns, feature_shapes,
+                                 feature_types):
+        column = df[col].values
+        if column.dtype == np.object:
+            if isinstance(column[0], np.ndarray):
+                column = np.stack(column)
+            elif isinstance(column[0], (list, tuple)):
+                column = list(column)
+            else:
+                raise Exception(
+                    f"Column {col}'s type: {type(column[0])} is not supported."
+                    " It must be numpy built in type or numpy object of "
+                    "(ndarray, list, tuple)")
+
+        t = torch.as_tensor(column, dtype=dtype)
+        if shape is not None:
+            t = t.view(*(-1, *shape))
+        feature_tensor.append(t)
+
+    label_df = df[label_column].values
+    label_tensor = torch.as_tensor(label_df, dtype=label_type)
+    if label_shape:
+        label_tensor = label_tensor.view(-1, label_shape)
+    return feature_tensor, label_tensor
+
+
+class TorchDataset:
```

Comment:
Is there an interface for this?

Reply:
no, is it needed? Do you mean inherit from `ray.util.sgd.data.Dataset`?

Reply:
Or inherit from a common interface between TorchDataset and TFDataset. It is hard to program against this API if there is no common interface, right?

Reply:
added